feat: hash contract class logs with poseidon + add preimage to blob

MirandaWood opened the pull request:
~Quick and dirty~ Somewhat clean implementation which:

- Converts our SHA based contract class `LogHash`es with poseidon hashes and constrain them in the registerer.
- Verifies this ^ hash in the base rollup and appends to the blob as fields, eliminating the need for any calldata.
- Treats the logs as fields in ts instead of bytes.

EDIT: Added log validation to tx in this commit: https://github.com/AztecProtocol/aztec-packages/pull/12061/commits/cb739a21aca3719ba75b41f7a68cd268bd09f448

EDIT 2: We may not need to constrain the hash in the registerer, so can revert to the oracle hashing the log and leaving the constraining to the base rollup in a future PR.

MirandaWood committed:
feat: hash cc log w poseidon, add raw log to rollup + ts

MirandaWood committed:
feat: final changes, fix hash, cleanup

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/cc-logs-pos-hash

MirandaWood committed:
chore: aaaa safety doc change again

MirandaWood committed:
chore: aaaa safety doc change again again

MirandaWood commented on noir-projects/noir-contracts/contracts/contract_class_registerer_contract/src/main.nr:
```diff
@@ -217,13 +219,19 @@ pub contract ContractClassRegisterer {
+        let log_to_emit: [Field; CONTRACT_CLASS_LOG_SIZE_IN_FIELDS] =
+            array_concat(log, [0; CONTRACT_CLASS_LOG_SIZE_IN_FIELDS - N]);
+        // TODO(MW): use chunked poseidon if cheaper? Use variable poseidon with length of non-0 fields?
         let log_hash =
-            unsafe { emit_contract_class_unencrypted_log_private(contract_address, log, counter) };
+            std::hash::poseidon2::Poseidon2::hash(log_to_emit, CONTRACT_CLASS_LOG_SIZE_IN_FIELDS);
+        // Safety: TODO(MW): constraining below length in the base rollup. Is this good/safe?
+        let length = unsafe { find_index_hint_from_end(log_to_emit, |elem| !is_empty(elem)) };
+        // Safety: the below only exists to broadcast the raw log, so we can provide it to the base rollup later to be constrained.
```
Two Qs here:
- I'm just hashing the entire (padded) log here because it's easy and clean. I could hash just the N input fields. This is a bit more expensive in the circuits (not by much, I'd use the chunk variable hash) but might be a lot faster elsewhere. I already have code to hash N of M fields, and check the remaining M - N fields are all zero.
- I'm carrying the log length (N) throughout the kernels to track gas since we are only adding N things to the blob. I'm not sure this is even required atm because logs are set to cost 0? I'm also not constraining this length until the base rollup when we verify the hash. I'm not sure if this is ok.

iAmMichaelConnor replied:
I'd potentially go with the unpadded approach, since our recent discussions (via that hackmd) suggest that eventually (probably in a separate PR), we don't actually need to hash in this circuit at all.

We probably do need to track the length, because ultimately these logs should cost a relative fortune. And eventually (as per prev paragraph) only the Base will need to constrain this length (because we'll remove constrained hashing of this log from private-land).

MirandaWood replied:
> because we'll remove constrained hashing of this log from private-land

Shall I just do this now, and make the oracle hash again? I don't remember now - but wasn't there some reason we weren't happy with the original oracle hash even though we verified the hash later?

iAmMichaelConnor replied:
Sorry for the delay; I need to sort out my notification settings.
> Shall I just do this now, and make the oracle hash again?

Can do! It can either be an oracle, or just make the function `unconstrained`? But you messaged 14 hours ago, so maybe you've already chosen an approach, which I'll find out when I read further down this page :)

iAmMichaelConnor replied:
> wasn't there some reason we weren't happy with the original oracle hash even though we verified the hash later?

I can't remember 😬

MirandaWood replied:
Just FYI Alvaro found the unpadded hash is insecure unless we also constrain the length in the registeter (which just adds gates. Fixed len hash is about the same gates). From @sirasistant:

> I think that opens the attack vector:
> - Private chooses a shorter length
> - Private only hashes the shorter length
> - Base rollup receives the shorter preimage
> - Hashes match
> - We publish incomplete bytecode

I willl revert to a fixed length hash

MirandaWood commented on yarn-project/archiver/src/archiver/kv_archiver_store/log_store.ts:
```diff
@@ -391,19 +417,20 @@ export class LogStore {
-      if (!filter.contractAddress || log.contractAddress.equals(filter.contractAddress)) {
-        if (log instanceof UnencryptedL2Log) {
-          results.push(new ExtendedUnencryptedL2Log(new LogId(blockNumber, txIndex, logIndex), log));
+      // TODO(MW): fix hack for contract class
+      if (!filter.contractAddress || (log instanceof PublicLog && log.contractAddress.equals(filter.contractAddress))) {
+        if (log instanceof ContractClassLog) {
+          results.push(new ExtendedContractClassLog(new LogId(blockNumber, txIndex, logIndex), log));
```
The contract class log here will be a siloed field array. Initially I thought there was no need to have a `.contractAddress` property on `ContractClassLog`s because by this point they would have been constrained to be from the registerer (regardless of whether they contain a valid event).
However this totally bricked a lot of tests. I added a quick hack that accepts any cc log to get them running. But maybe we should carry a contract address around with the log?

iAmMichaelConnor replied:
> maybe we should carry a contract address around with the log?

This sounds sensible!

MirandaWood commented on yarn-project/archiver/src/archiver/memory_archiver_store/memory_archiver_store.ts:
```diff
@@ -703,34 +712,44 @@ export class MemoryArchiverStore implements ArchiverDataStore {
-              }
+        for (let logIndex = 0; logIndex < blockLogs.length; logIndex++) {
+          const log = blockLogs[logIndex];
+          // TODO(MW): this is a hack
+          const thisContractAddress = [
+            REGISTERER_CONTRACT_CLASS_REGISTERED_TAG,
+            REGISTERER_PRIVATE_FUNCTION_BROADCASTED_TAG,
+            REGISTERER_UNCONSTRAINED_FUNCTION_BROADCASTED_TAG,
+          ].includes(log.fields[0])
+            ? AztecAddress.fromNumber(REGISTERER_CONTRACT_ADDRESS)
+            : AztecAddress.ZERO;
```
Similar to the comment on `log_store`, since we don't carry a `.contractAddress` around with the cc log, it's tricky once we get to the archiver to allow for filtering by address.

MirandaWood committed:
feat: add cc log validation to tx

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/cc-logs-pos-hash

MirandaWood committed:
chore: post merge fixes, fmt

MirandaWood committed:
chore: fixtures + more import fixing

MirandaWood committed:
feat: use unpadded variable hash for cc logs

MirandaWood committed:
fix: update tests

MirandaWood committed:
feat: add .contractaddr to contract class log (#12197)

Tracking an address alongside the cc log to aid filters in the archiver.
The address is always constrained to be that of the registerer, but many
archiver tests use random addresses so they brick without ugly hacks

MirandaWood committed:
chore: move cc log to /logs/ to prep for conflicts

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/cc-logs-pos-hash

MirandaWood committed:
chore: post merge fmt and fixes

MirandaWood committed:
chore: aaa lint

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/cc-logs-pos-hash

MirandaWood committed:
chore: post merge fmt

iAmMichaelConnor commented:
:eyes:

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/cc-logs-pos-hash

MirandaWood committed:
chore: post merge fixes, fmt

MirandaWood committed:
chore: remove docker fmt i guess

MirandaWood committed:
chore: more fixes (bootstrap no longer full build)

MirandaWood commented on yarn-project/end-to-end/scripts/docker-compose.yml:
Sorry - this file is changed by `yarn format` and I forgot to remove from staged - nothing to do with PR. If the next CI run passes, I'll remove it.

iAmMichaelConnor commented on yarn-project/stdlib/src/logs/contract_class_log.ts:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  static async random() {
+    // TODO(MW): Lazily used /2 instead of CONTRACT_CLASS_LOG_DATA_SIZE_IN_FIELDS, because this keeps overfilling block blobs
+    // Below line gives error 'Type instantiation is excessively deep and possibly infinite. ts(2589)'
+    // makeTuple(CONTRACT_CLASS_LOG_DATA_SIZE_IN_FIELDS, Fr.random);
+    const fields = Array.from({ length: Math.ceil(CONTRACT_CLASS_LOG_DATA_SIZE_IN_FIELDS / 2) }, () => Fr.random());
+    return new ContractClassLog(
+      await AztecAddress.random(),
+      fields.concat(Array.from({ length: Math.floor(CONTRACT_CLASS_LOG_DATA_SIZE_IN_FIELDS / 2) }, () => Fr.ZERO)),
+    );
```
Maybe comment something like "Only half of the contract class log's fields get filled with random values here (ideally all of the fields would be filled with random values), because...". It took my brain a sec to understand this.

iAmMichaelConnor replied:
Can we not ignore the error, like I saw you do elsewhere?

MirandaWood replied:
No - it's a nightmare because they are tuples, not arrays. I couldn't get tuples defined in our usual way working at all

iAmMichaelConnor commented on yarn-project/stdlib/src/logs/contract_class_log.ts:
```diff
@@ -0,0 +1,112 @@
+  }
+
+  async hash() {
+    return await poseidon2Hash(this.getEmittedFields());
```
Maybe comment that this is one of two approaches to hashing this data, to disambiguate from the weird AVM way of hashing the data.

iAmMichaelConnor commented on yarn-project/stdlib/src/tx/tx.ts:
```diff
@@ -170,6 +171,40 @@ export class Tx extends Gossipable {
 
+  /**
+   * Gets either revertible or non revertible contract class logs emitted by this tx.
+   * @param revertible - true for revertible only logs, false for non reverible only logs.
```
Oh gosh, I didn't even pay attention to any logic that discerned between rev and non-rev in the `.nr` files...

MirandaWood replied:
I didn't need to change it because before this PR we use a log hash in the kernels. After this PR we... use a log hash in the kernels :)

iAmMichaelConnor reviewed (commented):
Wooooh I read it all! Thanks for doing all this! It's quite scary how much of the codebase needs to be touched to change a very very thin "slice" of functionality. It's essentially just changing a public input of the private context, and it blows up into thousands of lines. 
Anyway, I've written comments, and a lot of them will be because this is the first time I'm looking at most of these files. :)

MirandaWood commented on noir-projects/noir-contracts/contracts/contract_class_registerer_contract/src/main.nr:
```diff
@@ -217,13 +220,18 @@ pub contract ContractClassRegisterer {
         // cannot prove non-registration. Therefore, it is possible that a malicious oracle might prevent sequencers
```
True, I was hoping the reviewer would give me some guidance on this. If we move the hash to the oracle / unconstrained as planned this comment is still correct isn't it? Shall I change it then change it back?

MirandaWood commented on noir-projects/noir-protocol-circuits/crates/types/src/utils/arrays.nr:
```diff
@@ -101,6 +120,29 @@ where
 
+// Routine which validates that zero values of an array form a contiguous region at the end, i.e.,
+// of the form: [*,*,*...,0,0,0,0] where * is any value (zeroes allowed).
+pub fn validate_trailing_zeroes<T, let N: u32>(array: [T; N]) -> u32
```
Just to be clear, everything you commented on here existed before my PR. It's been part of the code for a while now to only validate an array once to avoid wasting gates validating it every time its used. A valid array never contains zeroes in the middle, it MUST be right padded.

Contract class logs are not one of these valid arrays and `validate_array` is never called it, only the trailing zeroes method.

iAmMichaelConnor replied:
Sure, yeah I opened an issue because I don't mean to grow the scope of this PR.
I'm just putting my security hat on, and that led me to suggestions like those in the issue I opened. Wrapping a validated array in a type that specifically conveys that it's been validated is better than trying to mentally track whether it's been validated yet.

MirandaWood commented on yarn-project/noir-protocol-circuits-types/src/conversion/server.ts:
```diff
@@ -292,6 +293,23 @@ export function mapBlockBlobPublicInputsFromNoir(
+export function mapContractClassLogToNoir(log: ContractClassLog): ContractClassLogNoir {
+  return {
+    log: {
+      // @ts-expect-error - below line gives error 'Type instantiation is excessively deep and possibly infinite. ts(2589)'
```
This is typescript with large arrays. This is already an issue throughout the codebase - I don't know enough about ts to say why it's a problem

MirandaWood commented on noir-projects/noir-protocol-circuits/crates/types/src/utils/arrays.nr:
```diff
@@ -230,6 +272,30 @@ fn smoke_validate_array() {
+    let valid_array = [1, 0, 3, 0];
+    assert(validate_trailing_zeroes(valid_array) == 3);
+
+    let valid_array = [1, 2, 3, 0, 0];
+    assert(validate_trailing_zeroes(valid_array) == 3);
```
I believe these are all tested and handled inside `nr` tests, either at the bottom of this file or in dedicated `/test/` directories

MirandaWood commented on noir-projects/noir-protocol-circuits/crates/private-kernel-lib/src/components/tail_output_composer/meter_gas_used.nr:
```diff
@@ -31,9 +31,16 @@ pub fn meter_gas_used(data: PrivateToRollupAccumulatedData) -> Gas {
+    let num_contract_class_logs = array_length(data.contract_class_logs_hashes);
+    let contract_class_log_emitted_fields = data.contract_class_logs_hashes.fold(
+        0,
+        // TODO(MW): log_size is constrained to be correct in the base rollup
```
This is related to a comment elsewhere, because I'm not sure constraining in base is sufficient. I think based on your review it is? So I can remove this

MirandaWood commented on noir-projects/noir-protocol-circuits/crates/types/src/abis/contract_class_log.nr:
```diff
@@ -0,0 +1,44 @@
+    }
+}
+
+pub fn validate_and_silo_contract_class_log(
```
I can move it to `components` but I don't think splitting it will help. It would be confusing I think for readers when all we need to do is prepare the logs for broadcast, which means check the length and silo them. Maybe a rename would work?

MirandaWood commented on noir-projects/noir-protocol-circuits/crates/rollup-lib/src/base/public_base_rollup.nr:
```diff
@@ -38,7 +42,7 @@ impl PublicBaseRollupInputs {
         reverted: bool,
         combined_constant_data: CombinedConstantData,
-    ) -> TxEffect {
+    ) -> (TxEffect, [u32; MAX_CONTRACT_CLASS_LOGS_PER_TX]) {
```
Yes, exactly. No other effects have a length claim because they are raw from the kernels. If we had raw contract class logs, like the previous PR, this wouldn't be needed

iAmMichaelConnor commented on noir-projects/noir-protocol-circuits/crates/types/src/utils/arrays.nr:
```diff
@@ -81,6 +81,25 @@ pub unconstrained fn find_index_hint<T, let N: u32, Env>(
+        let j = N - i - 1;
+        // We check `index == 0` to ensure that we only update the index if we haven't found a match yet.
+        if (index == 0) & find(array[j]) {
+            // Since we are looking from the end of the array, the index at which find() is false is the next one:
+            index = j + 1;
```
Oh, sorry i meant to say `find_index_hint_from_end` _should_ return `4` for array `[0, 1, 2, 3, 4, 0, 0]`, and its length should be calculated separately as 5.

MirandaWood replied:
I think I had that originally (in the original logs as fields PR) but will change it to this ^ - makes sense

MirandaWood commented on yarn-project/archiver/src/archiver/data_retrieval.ts:
```diff
@@ -258,22 +254,10 @@ async function getBlockFromRollupTx(
+  const blockBody = Body.fromBlobFields(blockFields);
 
-  // TODO(#9101): Once we stop publishing calldata, we will still need the blobCheck below to ensure that the block we are building does correspond to the blob fields
+  // TODO: Will this ever throw now that we do not get blocks from calldata at all?
```
In master, it's getting some calldata (just these logs) to be able to reconstruct a block (our `L2Block`s) from L1 info. In this PR, block calldata does not exist - all information we need to construct a block exists in a blob. There is some calldata do with other stuff (signatures etc?) that I don't know much about, so I didn't rename the function from `extractRollupProposeCalldata` to something like `extractRollupProposeData`.

My edit to this comment is wondering whether we are actually doing any comparison here at all. With this PR we get zero block data from `extractRollupProposeCalldata` -  it's all from the blob. So there is one source and it can't get of sync. But I also wanted to be sure, hence the comment

LeilaWang replied:
When reviewing the first huge blob pr we had a discussion around this and agreed that we could remove this redundant check when calldata is not used anymore!

MirandaWood replied:
Thank you, had forgotten! I will include this in a follow up

MirandaWood commented on yarn-project/archiver/src/archiver/memory_archiver_store/memory_archiver_store.ts:
```diff
@@ -698,34 +699,36 @@ export class MemoryArchiverStore implements ArchiverDataStore {
     const contractAddress = filter.contractAddress;
 
-    const logs: ExtendedUnencryptedL2Log[] = [];
+    const logs: ExtendedContractClassLog[] = [];
```
I believe so yes

MirandaWood commented on yarn-project/noir-protocol-circuits-types/src/conversion/client.ts:
```diff
@@ -252,56 +251,6 @@ function mapScopedPrivateLogDataFromNoir(data: ScopedPrivateLogDataNoir) {
   );
 }
 
-/**
```
Yep, either duplicates of existing methods or completely unused methods

MirandaWood commented on noir-projects/noir-protocol-circuits/crates/rollup-lib/src/components.nr:
```diff
@@ -356,23 +358,47 @@ fn get_tx_effects_hash_input(
-    // Currently appending the single log hash as an interim solution
     // CONTRACT CLASS LOGS
-    array_len = array_length(contract_class_logs);
+    // Note: we can use contract_class_logs_lengths here as they are constrained to be zero if
+    // the log is empty, and the logs are constrained to be sorted L to R.
```
I think ALL `array_length` checks here rely on the arrays being somehow constrained previously, we should probably make that more clear everywhere (another PR?)

MirandaWood committed:
feat: address some comments

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/cc-logs-pos-hash

MirandaWood committed:
chore: post merge fixes

LeilaWang reviewed (approved):
Agree with some of Mike's comments, but maybe they can be addressed in followup prs. It looks great to me!

MirandaWood committed:
Merge remote-tracking branch 'origin/master' into mw/cc-logs-pos-hash

LeilaWang committed:
Merge remote-tracking branch 'origin/master' into mw/cc-logs-pos-hash

MirandaWood committed:
chore: fix merges and empty array lens

MirandaWood committed:
fix: i am stupid
